README
Most recent change:	  29-Sep-2020

The file Raster3D_3.0-7.tar.gz contains the full source distribution for
version 3.0 of Raster3D.

  See CHANGELOG for a history of changes to the package.
  Basically nothing has changed since verison 3.0-3 (2014) except for
  minor tweaks to accommodate newer Fortran versions.

  Note: The programs from Raster3D are included in the CCP4 crystallography
  suite.  You might find it easier to use that pre-built version.

- In a nutshell, the installation procedure is:

  	1)  Review the contents of Makefile.template
	2)  make help            (prints a list of OS names)
	3)  make OS              (where OS is your operating system name)
	4)  make all		 (builds the programs)
	5)  make install         (copies files to system directories)
	6)  modify user login environment to include R3D_LIB 
		e.g. setenv R3D_LIB /usr/local/share/Raster3D/materials

	7)  (optional)
	    modify ImageMagick delegates file so that ImageMagick recognizes
	    *.r3d files by default.  On my systems this file is in
	      /etc/ImageMagick-6/config/delegates.xml
	    Add the following line
	      <delegate decode="r3d" command="&quot;render&quot; &lt; &quot;%i&quot; &gt; &quot;%o&quot;"/>


- Changes in Version 3.0
	+ new code in rastep to support Skittles validation (introduced in 2.9)
	+ dynamic array allocation
	+ png/jpeg output using libgd
	+ labels handled by libgd

If you run into installation problems, please let me know,
particularly if you can also tell me what you did to fix it!

For optimal use of Raster3D you should also obtain 
- John Cristy's ImageMagick package for image manipulation and viewing 
	(source distribution available from www.ImageMagick.org)
- Version 2.0 or greater of Per Kraulis's program Molscript

To build support for direct output of TIFF files
- Some version of a TIFF library

License
-------
Raster3D Version 3 is distributed under the terms of the Artistic License,
Version 2.0.  A copy of the license is included with the source code.

				Ethan A Merritt
				merritt@u.washington.edu


CHANGELOG
29-Sep-2020     Package 3.0-7
13-May-2020     -std=legacy seems to reduce pedantry from gfortran 10.1
10-May-2020     gfortran 10.1 is insanely picky - must wrap hex constants in INT(...)
08-May-2020     Package 3.0-6
08-May-2020     Change flag for hexadecimal constant from X'000 to Z'000
08-May-2020     recent gfortran has inverted the meaning of compiler flag -Wtabs (?!)
12-Nov-2019     Package 3.0-5
12-Nov-2019	rods.f:  gfortran 8.3 requires decimal point after floating 0
12-Nov-2019	import into git
21-Jan-2019	Package 3.0-4
21-Jan-2019	suppress compiler warnings for unused variables
29-Mar-2014	Package (somewhat belatedly) 3.0-3
01-May-2012	local.c: only #include <in.h> if NETWORK_BYTEORDER
		Makefile: but don't define NETWORK_BYTEORDER for linux
12-Apr-2011	Default to rendering both sides of an opaque triangle with normals
		Ignore blank lines while looking for next object
27-Mar-2011	Remove obsolete directory of patch files for molscript v1.4
		Add notices to the man page, the html page, the spec files,
		and README that the program is distributed under the 
		Artistic License, Version 2.0
03-Feb-2011	Package 3.0-2 (SBGRID builds)
26-Jan-2011	Joe Krahn: use AlphaBlending mode 0 when rendering labels
19-Dec-2011	Package 3.0-1 (SBGRID builds)
18-Dec-2010	make NAX,NAY integer*4 to prevent overflow during autotiling
16-Dec-2010	modify stereo3d script to use render -labels rather than label3d
		hardwire libraries in LDFLAGS for building static rpm
		read in full header before initializing output
16-Dec-2010	Joe Krahn:
		Add Makefile entries for osx and osx-fink
		Use new Fortran intrinsic LEN_TRIM
		Choice of four transparency algorithms
15-Dec-2010	Handle UTF-8 sub and superscripts in label processing
		Use OPT(2) of MATERIAL record to select transparency algorithm
		Add a Makefile option for "make osx"
14-Dec-2010	Revise packaging instructions in Makefile.package
14-Dec-2010	add first cut of modified render.f local.c r3dtogd.f that use libgd
		to produce png/jpeg output and also to handle labels
14-Dec-2010	Add dummy MODULES to normal3d.f so that the symbols are defined for qinp
04-Apr-2010	updated the version on raster3d-3.0.spec and raster3d-static-3.0.spec SPEC files
30-Mar-2010	added the raster3d-static-2.9.spec SPEC file for static builds
16-Mar-2010	Release 2.9-2 (dynamic)
12-Feb-2010	Release 2.9-1 (static)
12-Feb-2010	added escape characters for dollar sign in $(FC) in Makefile
09-Feb-2010	expandable array allocations to for MAXOBJ, MAXN*XY, etc
09-Feb-2010	created Raster3D-3.0 branch from Raster3D-2.9
26-Jan-2010	bring ribbon.f ribbon1.f into the gfortran era
22-Jan-2010	commit my current source tree to cvs repository on verdandi
22-Jan-2010	avs2ps.c: more brackets in the initializer for avail[][]
		local.c:  add casts to quiet compiler warning
22-Dec-2009	render.f: initialize fullname character storage
03-Dec-2009	suv.f: CCuv() was failing to initialize VV(3,2)
22-Oct-2009	VERSION VERSION.incl: bump version to 2.9EAM
22-Oct-2009	rastep.f: -aniso flag to include isotropic atoms in the analysis
12-Oct-2009	suv.f:		additional checks for NPD matrices
02-Sep-2009	suv.f:		wrapper function CCuv
		rastep.f:	Calculate CCuij for C-N bonds
19-Aug-2009	rastep.f   slight modification to format of output from -tabulate
04-Nov-2008	rings3d.f: -ribose option
15-Aug-2008	rings3d.f: Initialize nhits array
		           Allow for left-shifted atom labels in input PDB file
13-Mar-2008	qinp fails if compiled with gfortran and -On (n > 0)
		so break out qinp.f into separate file for non-optimized compilation
07-Mar-2008
		Release 2.7s   (band-aid release to support recent gfortran)
07-Mar-2008	render.f Makefile local.c normal3d.f quadricf parse.f rastep.f
		yet more initializers to make gfortran happy
		static structures in local.c
		gfortran optimization level must be -O2 or -O3 (else bad code)
25-Feb-2008	render.f	 	Initialize NSXMAX, NSYMAX to zero
25-Feb-2008	VERSION VERSION.incl	Development version on skuld versioned as 2.7s
		Release 2.7d
 3-May-2006	rastep.f: gfortran requires initialization to zero!
12-Apr-2006	label3d: ImageMagick 6+ requires different syntax to set transparency
12-Apr-2006	r3dtops.f: gfortran requires SAVE, because it doesn't like -fno-automatic
11-Apr-2006	In example scripts, change "display -geometry" to "display -resize"
11-Apr-2006	change AND() and OR() to iand() and ior()
11-Apr-2006	rings3d.f: gfortran doesn't like iargc declared external
11-Apr-2006	avs2ps.c: additional headers (Debian patch)
11-Apr-2006	ungz.c: additional headers (Debian patch)
11-Apr-2006	local.c: additional headers (Debian patch)
15-Dec-2005	rastep.f: declare DEFCOLS as integer
15-Dec-2005	parse.f: and -> iand
15-Dec-2005	parse.f: hexadecimal syntax change
15-Dec-2005	render.f: and -> iand
15-Dec-2005	render.f: hexadecimal syntax change
15-Dec-2005	local.c: provide substitute or, and for gfortran
15-Dec-2005	Update Makefile for gfortran		
		Release 2.7c
12-Feb-2004	rastep.f: g77 doesn't like logical test against noerr
 6-Feb-2004	modify label3d and stereo3d scripts to deal with changes to recent
 		versions of ImageMagick and ghostscript
17-Oct-2003	rastep now checks input for illegal (=0) axial Uij terms
19-Aug-2003	render -bg {white|black|#RRGGBB}  set background from command line
		Release 2.7b
19-Jun-2003	Fix initialization of TMPNAM in normal3d.f 		
		Release 2.7a
 6-May-2003	render, and all other programs,  now default to PNG output.
		Add sun-forte Makefile option (Chandler Wilkerson <chandler@cs.uh.edu>)
 5-May-2003	Modify default color table in rastep so that if atom name column
 		is one off it still works.
 6-Nov-2002	Release 2.6f (mostly relevant only to RPM installation) 
 6-Nov-2002	Change default image format in scripts to PNG rather than TIFF
 		Modify RPM installation script for newer ImageMagick delegates format
		Build RPM against newer libpng.so.3 (Mandrake 8.2/9.0 Redhat 7.3/8.0)
26-Jul-2002	Suppress error message if libtiff has no LZW compression

parse.f
	subroutine parse
c	Version 2.7c
c
      COMMON /OPTIONS/ FONTSCALE, GAMMA, ZOOM, NSCHEME, SHADOWFLAG, XBG, 
     &                 NAX, NAY, OTMODE, QUALITY, INVERT, LFLAG
      REAL             FONTSCALE, GAMMA, ZOOM
      INTEGER          NSCHEME, SHADOWFLAG, XBG
      INTEGER*4        NAX, NAY, OTMODE, QUALITY
      LOGICAL*2        INVERT, LFLAG
c
	common /asscom/  assout, verbose
	integer          assout
	logical                  verbose
c
	character*132    args(4), option, labelfile
	save             args
c
	integer    INVERTFLAG
	parameter (INVERTFLAG=8)
c
	do larg = 1,4
	    args(larg) = ' '
	end do
	larg = 0
	verbose = .false.
	nax = -1
	nay = -1
	nscheme = -1
	shadowflag = -1
	xbg = 0
	quality = 90
	lflag = .false.
	zoom = 0.
	gamma = 1.0
c
c	Default font scale is 3.0 (appropriate for 300 dpi printers)
c	This is superseded by the environmental variable FONTSCALE
c	or a specific command line options -fontscale XX
	call getenv('FONTSCALE',option)
	read (option,*,err=8,end=8) fontscale 
	goto 9
    8	fontscale = 3.0
    9	continue
c
	iarg  = 1
	nargs = iargc()
   10	continue
	if (iarg .gt. nargs) goto 11
	    call getarg( iarg, option )
	    if (option(1:6).eq.'-debug') then
	    	verbose = .true.
	    else if (option(1:3).eq.'-aa') then
	    	if (option(4:4).eq.'0') then
		    nscheme = 0
		else if (option(4:4).eq.'1') then
		    nscheme = 1
		else
		    nscheme = 4
		end if
	    else if (option(1:6).eq.'-draft') then
	    	nscheme = 1
	    else if (option(1:6).eq.'-alpha') then
	    	nscheme = 0
	    else if (option(1:7).eq.'-transp') then
	    	nscheme = 0
	    else if (option(1:5).eq.'-qual') then
	    	iarg = iarg + 1
		call getarg( iarg, option )
		read (option,*,err=10,end=10) quality
		if (quality.le.0 .or. quality.gt.100) quality = 95
	    else if (option(1:5).eq.'-size') then
	    	iarg = iarg + 1
		call getarg( iarg, option )
		kbrk = 2
		do k = 15,2,-1
		    if (option(k:k).eq.'x') kbrk = k
		    if (option(k:k).eq.'X') kbrk = k
		end do
		read (option(1:kbrk-1),*,err=10,end=10) nax
		read (option(kbrk+1:15),*,err=10,end=10) nay
		if (nscheme.lt.0) nscheme = -4
	    else if (option(1:6).eq.'-label') then
	    	lflag = .true.
		call getarg( iarg+1, option )
		if (iarg.lt.nargs .and. option(1:1).ne.'-') then
		    labelfile = option
		    iarg = iarg + 1
		else
		    labelfile = 'label3d.ps'
		end if
	    else if (option(1:10).eq.'-fontscale') then
	        iarg = iarg + 1
		call getarg( iarg, option )
		read (option,*,err=10,end=10) fontscale
		if (fontscale.le.0) fontscale = 3.0
	    else if (option(1:10).eq.'-gamma') then
	        iarg = iarg + 1
		call getarg( iarg, option )
		read (option,*,err=10,end=10) gamma
		if (gamma.le.0) gamma = 1.0
	    else if (option(1:5).eq.'-zoom') then
	    	iarg = iarg + 1
		call getarg( iarg, option )
		kbrk = 15
		do k = 15,2,-1
		    if (option(k:k).eq.'%') kbrk = k
		end do
		read (option(1:kbrk-1),*,err=10,end=10) zoom
		if (zoom.le.0.) zoom = 0.
		if (option(kbrk:kbrk).eq.'%') zoom = -zoom
	    else if (option(1:7).eq.'-shadow') then
	    	shadowflag = 1
	    else if (option(1:9).eq.'-noshadow') then
	    	shadowflag = 0
	    else if (option(1:3).eq.'-bg') then
	    	iarg = iarg + 1
		call getarg( iarg, option )
		if (option(1:5).eq.'white') xbg = INT(Z'FFFFFF')
		if (option(1:5).eq.'black') xbg = INT(Z'000000')
		if (option(1:1).eq.'#') then
		    read  (option(2:7),'(1Z6)') xbg
		    if (verbose) then
		        write (0,'(1Z12)') xbg
		    endif
		endif
	    else 
	    	larg = larg + 1
		args(larg) = option
	    end if
	iarg = iarg + 1
	goto 10
   11	continue
c
	if (lflag) call lopen(labelfile)
c
	otmode = local(0, args(1), args(2), args(3), args(4))
c
	if (iand(otmode,invertflag).ne.0) then
	    invert = .false.
	else
	    invert = .true.
	end if
c
	end

ccc
cc
c
	subroutine autotile( nax, nay, modulus )
c
	implicit NONE
c
	include 'parameters.incl'
c
	common /raster/  ntx,nty,npx,npy
	integer          ntx,nty,npx,npy
c
	common /asscom/  assout, verbose
	integer          assout
	logical                  verbose
c
	integer*4 nax, nay
c
c	Anti-aliasing requires that NPY is a multiple of modulus
	integer   modulus
c
	if (nax .gt. 0) then
	    npx = modulus
	    ntx = (nax+npx-1) / npx
   21	    continue
	    if (mod(ntx,2).eq.0) then
	    	ntx = ntx / 2
		npx = npx * 2
	    else if (mod(ntx,3).eq.0) then
	    	ntx = ntx / 3
		npx = npx * 3
	    else if (mod(ntx,5).eq.0) then
	    	ntx = ntx / 5
		npx = npx * 5
	    else
	    	ntx = ntx + 1
	    end if
	    if (ntx.gt.MAXNTX) goto 21
	    if (npx.lt.6 .and. ( mod(ntx,2).eq.0
     &	        .or. mod(ntx,3).eq.0 .or. mod(ntx,5).eq.0))
     &		goto 21
	    if (npx.gt.MAXNPX) then
		npx = (MAXNPX/modulus) * modulus
		ntx = (nax + npx - 1) / npx
	    endif
	endif
	if (nay .gt. 0) then
	    npy = modulus
	    nty = (nay+npy-1) / npy
   31	    continue
	    if (mod(nty,2).eq.0) then
	    	nty = nty / 2
		npy = npy * 2
	    else if (mod(nty,3).eq.0) then
	    	nty = nty / 3
		npy = npy * 3
	    else if (mod(nty,5).eq.0) then
	    	nty = nty / 5
		npy = npy * 5
	    else
	    	nty = nty + 1
	    end if
	    if (nty.gt.MAXNTY) goto 31
	    if (npy.lt.6 .and. ( mod(nty,2).eq.0
     &	        .or. mod(nty,3).eq.0 .or. mod(nty,5).eq.0))
     &		goto 31
	    if (npy.gt.MAXNPY) then
		npy = (MAXNPY/modulus) * modulus
		nty = (nay + npy - 1) / npy
	    endif
	endif
	if (verbose .and. (nax.gt.0 .or. nay.gt.0)) then
	    write(0,32) 'X',ntx,npx,ntx*npx,nax
	    write(0,32) 'Y',nty,npy,nty*npy,nay
   32	    format('Autotiling on ',A1,': ',i5,' x ',i5,' = ',i7,
     &             ' >= ',i7)
	end if
c
	end
